modules/evm/src/lib.rs

```diff
@@ -467,21 +618,312 @@ impl<T: Trait> Module<T> {
+				.contract_info
+				.map_or(Err(Error::<T>::ContractNotFound), |_| Ok(()))
+		})?;
+		ensure!(
+			PendingTransferMaintainers::<T>::get(contract, new_maintainer).is_none(),
+			Error::<T>::PendingTransferMaintainersExists
+		);
+
+		let transfer_maintainer_deposit = T::TransferMaintainerDeposit::get();
+		let to = T::AddressMapping::to_account(&contract);
+		T::Currency::transfer(&who, &to, transfer_maintainer_deposit, ExistenceRequirement::KeepAlive)?;
+		T::Currency::reserve(&to, transfer_maintainer_deposit)?;
+		PendingTransferMaintainers::<T>::insert(contract, new_maintainer, transfer_maintainer_deposit);
+		Ok(())
+	}
+
+	fn do_confirm_transfer_maintainer(who: T::AccountId, contract: H160, new_maintainer: H160) -> DispatchResult {
+		PendingTransferMaintainers::<T>::mutate_exists(
+			contract,
+			new_maintainer,
+			|maybe_transfer_maintainer_deposit| -> DispatchResult {
+				let transfer_maintainer_deposit = maybe_transfer_maintainer_deposit
+					.as_mut()
+					.ok_or(Error::<T>::PendingTransferMaintainersNotExists)?;
+
+				Accounts::<T>::mutate(contract, |maybe_account_info| -> DispatchResult {
+					let account_info = maybe_account_info.as_mut().ok_or(Error::<T>::ContractNotFound)?;
+					let contract_info = account_info
+						.contract_info
+						.as_mut()
+						.ok_or(Error::<T>::ContractNotFound)?;
+
+					ensure!(who == contract_info.maintainer, Error::<T>::NoPermission);
+
+					let maintainer = T::AddressMapping::to_account(&new_maintainer);
+
+					let from = T::AddressMapping::to_account(&contract);
+					ensure!(
+						T::Currency::unreserve(&from, *transfer_maintainer_deposit).is_zero(),
+						Error::<T>::UnreserveFailed
+					);
+					T::Currency::transfer(
+						&from,
+						&contract_info.maintainer,
+						*transfer_maintainer_deposit,
+						ExistenceRequirement::AllowDeath,
+					)?;
+
+					contract_info.maintainer = maintainer;
+					Ok(())
+				})?;
+
+				*maybe_transfer_maintainer_deposit = None;
+				Ok(())
+			},
+		)
+	}
+
+	fn do_reject_transfer_maintainer(who: T::AccountId, contract: H160, invalid_maintainer: H160) -> DispatchResult {
+		PendingTransferMaintainers::<T>::mutate_exists(
+			contract,
+			invalid_maintainer,
+			|maybe_transfer_maintainer_deposit| -> DispatchResult {
+				let transfer_maintainer_deposit = maybe_transfer_maintainer_deposit
+					.as_mut()
+					.ok_or(Error::<T>::PendingTransferMaintainersNotExists)?;
+
+				Accounts::<T>::get(contract).map_or(Err(Error::<T>::ContractNotFound), |account_info| {
+					account_info
+						.contract_info
+						.map_or(Err(Error::<T>::ContractNotFound), |contract_info| {
+							if who != contract_info.maintainer {
+								Err(Error::<T>::NoPermission)
+							} else {
+								Ok(())
+							}
+						})
+				})?;
+
+				let from = T::AddressMapping::to_account(&contract);
+				let to = T::AddressMapping::to_account(&invalid_maintainer);
+
+				ensure!(
+					T::Currency::unreserve(&from, *transfer_maintainer_deposit).is_zero(),
```

Comment:
this should repatriate_reserved the reserve from requester to contract maintainer